assets/src/loader.rs

```diff
@@ -0,0 +1,180 @@
+              N: Into<String>,
+              S: Store + Send + Sync + 'static,
+              S::Error: Send
+    {
+        let context = self.contexts
+            .get(&TypeId::of::<A>())
+            .expect("Assets need to be registered");
+
+        let context: &Arc<A::Context> = context.downcast_ref().unwrap();
+
+        load_asset_future::<A, F, N, S>(context.clone(), format, name, store, &*self.pool)
+    }
+}
+
+/// Loads an asset with a given context, format, specifier and storage right now.
+pub fn load_asset<A, F, N, S>(context: &A::Context,
+                              format: &F,
+                              name: N,
+                              storage: &S)
+                              -> Result<A, AssetError<A::Error, F::Error, S::Error>>
+    where A: Asset,
+          F: Format<Data = A::Data>,
+          N: Into<String>,
+          S: Store
+{
+    let name = name.into();
+
+    let store_id = storage.store_id();
+    let spec = AssetSpec::new(name.clone(), F::extension(), store_id);
+
+    A::cached(context, &spec)
+        .map(|a| Ok(a))
+        .unwrap_or_else(move || load_asset_inner(context, format, spec, storage))
+        .map_err(|e| AssetError::new(AssetSpec::new(name, F::extension(), store_id), e))
+}
+
+fn load_asset_inner<A, F, S>(context: &A::Context,
+                             format: &F,
+                             spec: AssetSpec,
+                             storage: &S)
+                             -> Result<A, LoadError<A::Error, F::Error, S::Error>>
+    where A: Asset,
+          F: Format<Data = A::Data>,
+          S: Store
+{
+    let bytes = storage
+        .load(A::category(), &spec.name, spec.ext)
+        .map_err(LoadError::StorageError)?;
+    let data = format.import(bytes).map_err(LoadError::FormatError)?;
+    let a = Asset::from_data(data, context)
+        .map_err(LoadError::AssetError)?;
+
+    A::asset_loaded(context, spec, &a);
+
+    Ok(a)
+}
+
+/// Like `load_asset`, but loads the asset on a worker thread and returns
+/// an `AssetFuture` immediately.
+pub fn load_asset_future<A, F, N, S>(context: Arc<A::Context>,
+                                     format: F,
+                                     name: N,
+                                     storage: Arc<S>,
+                                     thread_pool: &ThreadPool)
+                                     -> AssetFuture<A, AssetError<A::Error, F::Error, S::Error>>
+    where A: Asset + Send + 'static,
+          A::Context: Send + Sync + 'static,
+          A::Error: Send + 'static,
+          F: Format<Data = A::Data> + Send + 'static,
+          F::Error: Send + 'static,
+          N: Into<String>,
+          S: Store + Send + Sync + 'static,
+          S::Error: Send + 'static
+{
+    use futures::future::lazy;
+
+    let name = name.into();
+
+    let closure = move || load_asset::<A, F, _, S>(&*context, &format, name, &*storage);
+
+    AssetFuture(thread_pool.spawn_future_async(lazy(closure)))
```

Comment:
Interesting idea..I have to think about it. Just keep in mind that hot-reloading is more a feature for development, so it's not necessary for the final game.